### PR TITLE
inspector: successful connection and execution

### PR DESCRIPTION
**Pull request by kfarnung**
##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
inspector

##### Description
* Now runs without crashing for a simple script.
* Console messages are plumbed back to the debugger.

### CODE REVIEW

**Comment by kfarnung** [comment]
@digitalinfinity, @kunalspathak - can you take a quick look when you get a chance?

File: deps/chakrashim/src/base/platform/platform.cc [hunk 1]
```diff
@@ -35,7 +35,11 @@ void OS::VPrintError(const char* format, va_list args) {
 
 
 int OS::SNPrintF(char* str, int length, const char* format, ...) {
-  return 0;
+  va_list args;
+  va_start(args, format);
+  int result = _vsnprintf(str, length, format, args);
```
Comment (kfarnung) [inline deps/chakrashim/src/base/platform/platform.cc:40]: I need to fix this to build on Linux.

File: deps/chakrashim/src/v8string.cc [hunk 2]
```diff
@@ -91,8 +91,9 @@ int String::Write(uint16_t *buffer, int start, int length, int options) const {
   size_t count = 0;
   if (JsCopyStringUtf16((JsValueRef)this, start, length,
                          buffer, &count) == JsNoError) {
-    if (!(options & String::NO_NULL_TERMINATION)) {
-      buffer[count] = 0;
+    if (!(options & String::NO_NULL_TERMINATION) &&
+        (length == -1 || count < length)) {
```
Comment (kfarnung) [inline deps/chakrashim/src/v8string.cc:95]: This is a bug in xplat as well, but luckily nobody seems to hit it.  This matches the behavior of the V8 code, but potentially leaves a buffer unterminated if there's not enough room in the buffer to fit the string and the null-terminator.  My preference would likely be to force a null in at the end unless null termination is off, but that would require changing more code and potentially break in the future.

File: src/inspector_agent.cc [hunk 3]
```diff
@@ -72,15 +72,23 @@ std::string StringViewToUtf8(const StringView& view) {
...
+  const size_t length = view.length();
+  std::string result;
+
+  for (size_t i = 0; i < length; i++) {
+    uint16_t wCh = source[i];
+    assert((wCh & 0xFF80) == 0);
+
+    result.append(1, static_cast<char>(wCh));
+  }
+
```
Comment (kfarnung) [inline src/inspector_agent.cc]: This is obviously a hack and will assert if the assumption is broken.  Once we have ICU support it should be trivial to fix, but even if we don't we'll have to revisit.

File: deps/chakrashim/lib/chakra_inspector.js [hunk 4]
```diff
@@ -0,0 +1,271 @@
+// Copyright Microsoft. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the 'Software'), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and / or
```
Comment (kunalspathak) [inline deps/chakrashim/lib/chakra_inspector.js:6]: run jslint on this file. Pretty sure that are some issues.

File: deps/chakrashim/lib/chakra_inspector.js [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    if (this.type == Debug.ScriptBreakPointType.ScriptId) {
+      return 'scriptId';
+    } else if (this.type == Debug.ScriptBreakPointType.ScriptName) {
+      return 'scriptName';
+    } else if (this.type == Debug.ScriptBreakPointType.ScriptRegExp) {
+      return 'scriptRegExp';
+    }
+  };
+
+  V8Breakpoint.prototype.Set = function() {
```
Comment (kunalspathak) [inline deps/chakrashim/lib/chakra_inspector.js:209]: >reakpoint.prototype.Se [](start = 5, length = 22)

I am a bit confused. Why does this file look so similar to `chakra_debug.js`?

File: deps/chakrashim/src/v8context.cc [hunk 6]
```diff
@@ -79,10 +80,15 @@ Local<Context> Context::New(Isolate* external_isolate,
...
-  if (jsrt::Debugger::IsDebugEnabled() && isoShim->debugContext == nullptr) {
+  bool debugEnabled = jsrt::Debugger::IsDebugEnabled();
+  bool inspectorEnabled = jsrt::Inspector::IsInspectorEnabled();
+
+  if ((debugEnabled || inspectorEnabled) &&
+      isoShim->debugContext == nullptr) {
     // If JavaScript debugging APIs need to be exposed then make sure
-    // debugContext is available and chakra_debug.js is compiled. Inject
-    // v8debug object from chakra_debug.js in this context global object
+    // debugContext is available and chakra_debug.js/chakra_inspector.js are
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8context.cc]: >/ [](start = 52, length = 1)

nit: say "or" will be more clear.

File: deps/chakrashim/src/v8isolate.cc [hunk 7]
```diff
@@ -136,7 +136,6 @@ void Isolate::SetCaptureStackTraceForUncaughtExceptions(
   //
   // kpathak: you might want to look into mechanism of captureStackTrace in
   // chakra_shim.js.
-  CHAKRA_ASSERT(false);
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8isolate.cc:139]: >CHAKRA_ASSERT [](start = 2, length = 13)

why is this removed?

File: src/inspector_agent.cc [hunk 8]
```diff
@@ -72,15 +72,23 @@ std::string StringViewToUtf8(const StringView& view) {
...
   ////const UChar* unicodeSource = reinterpret_cast<const UChar*>(source);
   ////static_assert(sizeof(*source) == sizeof(*unicodeSource),
   ////              "sizeof(*source) == sizeof(*unicodeSource)");
 
-  size_t result_length = view.length() * sizeof(*source);
-  std::string result(result_length, '\0');
```
Comment (kunalspathak) [inline src/inspector_agent.cc:83]: nit: why not delete this code?

File: deps/chakrashim/src/v8string.cc (hunk 2 above)
Comment (kunalspathak) [reply deps/chakrashim/src/v8string.cc:95]: /cc @jianchun

**Comment by kunalspathak** [comment]
LGTM overall except I am confused on why we can't reuse `chakra_debug.js` instead of having `chakra_inspector.js`.

File: deps/chakrashim/lib/chakra_inspector.js (hunk 4 above)
Comment (kfarnung) [reply deps/chakrashim/lib/chakra_inspector.js:6]: I don't see any issues reported.  There are a lot of issues with "vcbuild chakracore jslint" after Mark's changes merged, but I don't see any in this file.

Running "node .\tools\jslint.js .\deps\chakrashim\lib\chakra_inspector.js" just complains about ever line ending being a CRLF, no other issues.

File: deps/chakrashim/lib/chakra_inspector.js (hunk 5 above)
Comment (kfarnung) [reply deps/chakrashim/lib/chakra_inspector.js:209]: It's based upon it.  The Debug.setBreakPoint function needs to be present in the Debug context in order for module.js to use it.  This is here to keep that functionality working, but I'm not planning to implement everything in JS like last time.

File: deps/chakrashim/src/v8context.cc (hunk 6 above)
Comment (kfarnung) [reply deps/chakrashim/src/v8context.cc]: That makes sense, will fix.

File: deps/chakrashim/src/v8isolate.cc (hunk 7 above)
Comment (kfarnung) [reply deps/chakrashim/src/v8isolate.cc:139]: I wanted to unblock execution and will circle back.  For non-critical functions I'm leaving the CHAKRA-TODO, but removing the assert.

File: deps/chakrashim/lib/chakra_inspector.js (hunk 5 above)
Comment (kfarnung) [reply deps/chakrashim/lib/chakra_inspector.js:209]: The longer term plan is to delete the "debug" code (including chakra_debug.js) once that functionality is removed from upstream.

Comment (kunalspathak) [reply deps/chakrashim/lib/chakra_inspector.js:209]: Makes sense. Can you add a note in this file at the top stating these reasons, so some one new doesn't wonder the same thing?

File: src/inspector_agent.cc (hunk 8 above)
Comment (kfarnung) [reply src/inspector_agent.cc:83]: I can delete it, but the hope was that if we can enable ICU in the build that we would be able to just uncomment the code and enable Unicode support.

File: deps/chakrashim/src/v8string.cc (hunk 2 above)
Comment (kfarnung) [reply deps/chakrashim/src/v8string.cc:95]: I only found this because of the VC debug heap asserts that hit when the null terminator was placed just after the end of the buffer (the incoming parameters left exactly enough room for the characters).  I checked in the V8 code and they do basically the exact same check, skipping the null terminator in that case.
